walkthrough/step04.md

```diff
@@ -0,0 +1,217 @@
+# Step04 - Encoding cookies with JWT
+
+Now we have a user logging in and then setting the cookie, and from there we recognize the user through that cookie, but anyone can change that cookie. which is a big security vulnerability.
+
+
+`Token-based authentication` is a way to authenticate a user. The idea is that upon login we create a token (a string) and store it in a cookie. The browser then sends the token with subsequent requests (since cookies are sent with every request), and if the token can be verified (i.e. it is a legitimate token issued by the server), the user is authenticated.
+
+**would love to have an image here but couldn't find any good ones for tokens :smile: issues are welcomed**
+
+> Note: This token authentication is `stateless`, because no data persists on the server in relation to individual tokens. This is in contrast with stateful sessions, where a `session` is created in a database, and the session id is sent to the browser. Then the browser sends the id with each request, and the server checks in the database to validate the session. Being stateless, `token-based authentication` is less memory-intensive than `session-based authentication`, but both methods also have other pros and cons, which you can read about online.
+
+## Contents
+
+* [A more useful cookie](#a-more-useful-cookie)
+* [Signing](#signing)
+* [JSON Web Tokens](#json-web-tokens)
+* [exercise 1: creating and validating JWT cookies](#exercise-1-creating-and-validating-jwt-cookies)
+* [optional extra exercise 2- implement a HMAC](#optional-extra-exercise-2--implement-a-hmac)
+
+## A more useful cookie
+
+A very simple cookie (key and value) could look like `logged_in=true` or `username=druscilla` but sometimes we want to store more, and better-structured data.
+
+:star: How about JSON? :star:
+
+For example, inside a cookie, it might be useful to store:
+1. a user's id, and
+2. their access privileges.
+
+We could write in our handler:
+```javascript=
+const userInformation = {
+  userId: 45,
+  accessPrivileges: {
+    user: true,
+    admin: false
+  }
+};
+
+const cookieValue = JSON.stringify(userInformation);
+
+// Nodejs
+res.setHeader(
+  'Set-Cookie',
+  `data=${cookieValue}; HttpOnly; Secure`
+);
+
+// Express
+ res.cookie('data', cookieValue, {HttpOnly: true, secure: true});
+
+// the secure flag is to enable cookies only for
```

Comment:
I think this belongs to the previous step about cookies

Reply:
ohh yeah you're right there is a thorough explanation in step03